### DISCUSSION TITLE
How to avoid creating multiple sessions when using Dependencies with Security?

### DISCUSSION THREAD

**Discussion by DurandA** [fastapi/sqlmodel#555] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code ([full working example](https://github.com/DurandA/fastapi-security-example))

```python
from fastapi import Depends, Security
from fastapi.security import SecurityScopes
from sqlmodel import Session, create_engine, select


def get_session():
    with Session(engine) as session:
        yield session


async def get_current_user(
    security_scopes: SecurityScopes,
    token: str = Depends(oauth2_scheme),
    session: Session = Depends(get_session)
):
    # Token read omitted 👈
    user = session.get(User, token_data.user_uuid)
    if user is None:
        raise credentials_exception
    return user


async def get_current_app_user(
    current_user: User = Security(get_current_user, scopes=["app"])
):
    return current_user


@app.patch("/me", response_model=UserRead)
def update_current_user(
    user: UserUpdate,
    current_user: User = Depends(get_current_app_user).
    session: Session = Depends(get_session),
):
    user_data = user.dict(exclude_unset=True)
    for key, value in user_data.items():
        setattr(current_user, key, value)
    # Session and session from current_user are different ⚠️
    session.add(current_user) 
    session.commit()
    session.refresh(current_user)
    return current_user
```

### Description

I am using a synchronous (SQLite) session in a FastAPI Dependency as shown in the [SQLModel Tutorial](https://sqlmodel.tiangolo.com/tutorial/fastapi/session-with-dependency/#create-a-fastapi-dependency).

When a route "depends" on a `Dependency` and a `Security` that use `get_session`, the session is created twice resulting in errors such as:
```
sqlalchemy.exc.InvalidRequestError: Object '<User at 0x7f218cbf6a00>' is already attached to session '1' (this is '2')
```

How can this be prevented? I would expect a Security to [reuse Dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/sub-dependencies/#using-the-same-dependency-multiple-times) as long as `use_cache=True`.

### Operating System

Linux

### SQLModel Version

0.0.8

### Python Version

Python 3.8.10

I also created a [StackOverflow question](https://stackoverflow.com/questions/75487025/how-to-avoid-creating-multiple-sessions-when-using-fastapi-dependencies) about this issue.

**Comment by Magnaillusion** [comment]
You can use Python's functools to explicitly cache the dependency, like so:

```from fastapi import Depends, FastAPI, HTTPException, Query
from sqlmodel import Session, create_engine, select
from functools import cache


@cache
def get_session():
    with Session(engine) as session:
        yield session

# Omitting the rest of the code
```

This way, the session will be created only once. Any subsequent calls to the function (i.e. via dependencies) will return the cached object.

**Reply by DurandA** [reply]
Won't this prevent the session to be recreated for every request (which should be the normal behavior I believe)? I want the session to be created exactly once per request.

**Comment by DurandA** [comment, marked as answer]
The original code was oversimplified. The issue is introduced when there is both a `Security` object and a `Depends` that use `get_session()`. It looks more like a FastAPI bug with `Security` behavior than a SQLModel issue. Consequently I created a [FastAPI discussion](https://github.com/tiangolo/fastapi/discussions/6024) and mark this discussion as solved.
